File: Graph/DFS of Graph.py
'''
https://practice.geeksforgeeks.org/problems/depth-first-traversal-for-a-graph/1#
'''
#TC:O(V+E)
#SC:O(V)

    #Function to return a list containing the DFS traversal of the graph.
    def dfsOfGraph(self, V, adj):
        visited = [False]*V
        at = 0
        traversal = []
        
        def dfsHelper(visited,at,adj):
            nonlocal traversal
            if visited[at] :return 
            visited[at] = True
            traversal.append(at)
            for element in adj[at]:
                dfsHelper(visited,element,adj)
            return traversal
            
        return dfsHelper(visited,at,adj)

File: Graph/DFS of Graph.java
'''
https://practice.geeksforgeeks.org/problems/depth-first-traversal-for-a-graph/1#
'''
#TC:O(V+E)
#SC:O(V)

class Solution
{
    //Function to return a list containing the DFS traversal of the graph.
    public ArrayList<Integer> dfsOfGraph(int V, ArrayList<ArrayList<Integer>> adj)
    {
        boolean[] visited = new boolean[adj.size()];
        ArrayList<Integer> dfsList = new ArrayList<>();
        dfsHelper(0,visited,adj,dfsList);
        return dfsList;
    }
    public static void dfsHelper(int v,boolean[] visited,ArrayList<ArrayList<Integer>> adj,ArrayList<Integer> dfsList){
        visited[v] = true;
        dfsList.add(v);
        for(int adjacent:adj.get(v)){
            if(visited[adjacent]==false){
                dfsHelper(adjacent,visited,adj,dfsList);
            }
            
        }
    }
    }
